java/dataset/src/main/cpp/jni_wrapper.cc

```diff
@@ -27,7 +27,9 @@
 #include "arrow/dataset/file_base.h"
 #include "arrow/filesystem/localfs.h"
 #include "arrow/filesystem/path_util.h"
+#ifdef ARROW_S3
```

Comment:
Could you confirm whether `arrow/util/config.h` is included here implicitly?

Reply:
@kou, thanks for your review! I didn't find this header is included implicitly.

Reply:
OK. How about the following?

```diff
diff --git a/java/dataset/src/main/cpp/jni_wrapper.cc b/java/dataset/src/main/cpp/jni_wrapper.cc
index 19a43c8d2f..f1819aab0b 100644
--- a/java/dataset/src/main/cpp/jni_wrapper.cc
+++ b/java/dataset/src/main/cpp/jni_wrapper.cc
@@ -25,9 +25,8 @@
 #include "arrow/c/helpers.h"
 #include "arrow/dataset/api.h"
 #include "arrow/dataset/file_base.h"
-#include "arrow/filesystem/localfs.h"
+#include "arrow/filesystem/api.h"
 #include "arrow/filesystem/path_util.h"
-#include "arrow/filesystem/s3fs.h"
 #include "arrow/engine/substrait/util.h"
 #include "arrow/engine/substrait/serde.h"
 #include "arrow/engine/substrait/relation.h"
```

`arrow/filesystem/api.h` includes `arrow/util/config.h` explicitly and `ARROW_S3` must be defined if S3 is enabled with `arrow/filesystem/api.h`.

Reply:
@kou, by including [arrow/filesystem/api.h](https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/api.h), a few extra headers are directly or indirectly included, but they are not actually used here. Can we just keep the current change?

Reply:
If we want to keep the current style, we need to include `arrow/util/config.h` explicitly before we use `ARROW_S3`.

Reply:
@kou, I just realized what you meant. The code was updated. Thanks!